src/il_representations/envs/carla_envs.py

```diff
@@ -0,0 +1,121 @@
+import os
+import math
+import numpy as np
+import gym.spaces as spaces
+
+from il_representations.envs.config import (env_cfg_ingredient,
+                                            env_data_ingredient)
+from il_representations.envs.utils import stack_obs_oldest_first
+
+
+@env_data_ingredient.capture
+def _get_carla_data_opts(data_root, carla_demo_paths):
+    # workaround for Sacred issue #206
+    return data_root, carla_demo_paths
+
+
+@env_cfg_ingredient.capture
+def load_dataset_carla(task_name, carla_frame_stack, chans_first=True,
+                       n_traj=None):
+    data_root, carla_demo_paths = _get_carla_data_opts()
+
+    # load trajectories from disk
+    full_rollouts_path = os.path.join(data_root, carla_demo_paths[task_name])
+    trajectories = np.load(full_rollouts_path, allow_pickle=True)
+
+    obs = trajectories['observations'].astype(np.uint8)
+    nobs = trajectories['next_observations'].astype(np.uint8)
+    acts = trajectories['actions'].astype(np.uint8)
+    rews = trajectories['rewards'].astype(np.uint8)
+    dones = trajectories['terminals'].astype(np.uint8)
+
+    if n_traj is not None:
+        nth_traj_end_idx = [i for i, n in enumerate(cat_dones) if n][n_traj-1] + 1
+        obs = obs[:nth_traj_end_idx]
+        nobs = nobs[:nth_traj_end_idx]
+        acts = acts[:nth_traj_end_idx]
+        rews = rews[:nth_traj_end_idx]
+        dones = dones[:nth_traj_end_idx]
+
+    dataset_dict = {
+        'obs': obs,
+        'next_obs': nobs,
+        'acts': acts,
+        'rews': rews,
+        'dones': dones,
+    }
+
+    if chans_first:
+        for key in ('obs', 'next_obs'):
+            dataset_dict[key] = np.transpose(dataset_dict[key], (0, 3, 1, 2))
+
+    dataset_dict['obs'] = stack_obs_oldest_first(dataset_dict['obs'],
+                                                 carla_frame_stack,
+                                                 use_zeroed_frames=True)
+    dataset_dict['next_obs'] = stack_obs_oldest_first(dataset_dict['next_obs'],
+                                                 carla_frame_stack,
+                                                 use_zeroed_frames=True)
+
+    return dataset_dict
+
+class CarlaImageObsEnv:
```

Comment:
This should inherit from `gym.Wrapper` so that it gets all the standard env methods. Probably `gym.ObservationWrapper` is the ideal parent class from the perspective of minimising code required (you only have to override `__init__` to set the `observation_space`, then override `observation(x)` method to reshape observations).